### PR TITLE
[android][ios][ts][docs][ncl] implement background fetch on Android

### PR DESCRIPTION
**Pull request by tsapeta**
# Why

Closes https://expo.canny.io/feature-requests/p/android-background-fetch-support

# How

Android doesn't have its own background fetch feature, however I could easily imitate its behavior using `AlarmManager` with wakeup option.
I also had to change a few things in `TaskManager`, the most important one is the change in pending intents. Now, two pending intents created for the same task will be equal (see https://developer.android.com/reference/android/content/Intent#filterEquals(android.content.Intent)). Thanks to this, it's easier to cancel existing pending intents.

# Test Plan

Tested against `BackgroundFetch` example in `native-component-list`

### CODE REVIEW

**Comment by sjchmiela** [comment]
Sorry, we no longer accept new features in Expo.

**Comment by sjchmiela** [comment]
Just kidding! 🤣

File: docs/versions/unversioned/sdk/background-fetch.md [hunk 1]
```diff
@@ -17,13 +16,16 @@ Returns a promise resolving to one of these values:
...
+Registers background fetch task with given name. Registered tasks are saved in persistent storage and restored once the app is initialized.
+On iOS, it is also required to call [BackgroundFetch.setMinimumIntervalAsync](#backgroundfetchsetminimumintervalasyncminimuminterval) as fetch operations will not be occurring with its default value.
 
 #### Arguments
 
--   **taskName (_string_)** -- Name of the task to register. The task needs to be defined first - see [TaskManager.defineTask](../task-manager/#taskmanagerdefinetasktaskname-task) for more details.
+-   **taskName (_string_)** -- Name of the task to register. The task needs to be defined first - see [TaskManager.defineTask](../task-manager#taskmanagerdefinetasktaskname-task) for more details.
+-   **options (_object_)** -- An object of options:
+    -   **interval (_number_)** -- Inexact interval in seconds between subsequent repeats of the background fetch alarm. The final interval may differ from the specified one to minimize wakeups and battery usage. Defaults to **15 minutes**. (**Android only**, see [BackgroundFetch.setMinimumIntervalAsync](#backgroundfetchsetminimumintervalasyncminimuminterval) for iOS alternative)
```
Comment (sjchmiela) [inline docs/versions/unversioned/sdk/background-fetch.md]: Maybe on iOS we could call `BackgroundFetch.setMinimumIntervalAsync` behind the scenes when this option is provided? 🤔

File: packages/expo-background-fetch/src/BackgroundFetch.ts [hunk 2]
```diff
@@ -15,38 +16,39 @@ enum BackgroundFetchStatus {
...
 }
 
+interface BackgroundFetchOptions {
+  interval?: number,
+}
+
 export async function getStatusAsync(): Promise<BackgroundFetchStatus | null> {
   if (Platform.OS !== 'ios') {
-    return Promise.resolve(null);
+    return Promise.resolve(BackgroundFetchStatus.Available);
```
Comment (sjchmiela) [inline packages/expo-background-fetch/src/BackgroundFetch.ts]:
```suggestion
    return BackgroundFetchStatus.Available;
```

File: packages/expo-background-fetch/src/BackgroundFetch.ts [hunk 3]
```diff
@@ -15,38 +16,39 @@ enum BackgroundFetchStatus {
...
   }
   return ExpoBackgroundFetch.getStatusAsync();
 }
 
 export async function setMinimumIntervalAsync(minimumInterval: number): Promise<void> {
   if (Platform.OS !== 'ios') {
-    console.warn(`expo-background-fetch is currently available only on iOS`);
-    return;
+    return Promise.resolve();
```
Comment (sjchmiela) [inline packages/expo-background-fetch/src/BackgroundFetch.ts]:
```suggestion
    return;
```

**Review by sjchmiela — COMMENTED** [review]
Looks good to me, thanks! We could also try implementing starting service at the device's boot, couldn't we? I found this resource useful https://github.com/codepath/android_guides/wiki/Starting-Background-Services#using-with-alarmmanager-for-periodic-tasks

File: packages/expo-background-fetch/ios/EXBackgroundFetch/EXBackgroundFetchTaskConsumer.m [hunk 4]
```diff
@@ -20,6 +20,12 @@ - (NSString *)taskType
 - (void)didRegisterTask:(id<EXTaskInterface>)task
 {
   _task = task;
+  [self updateMinimumInterval];
+}
+
+- (void)setOptions:(NSDictionary *)options
+{
+  [self updateMinimumInterval];
```
Comment (sjchmiela) [inline packages/expo-background-fetch/ios/EXBackgroundFetch/EXBackgroundFetchTaskConsumer.m]: Just by looking at this diff, shouldn't we do something with these options?

File: packages/expo-background-fetch/ios/EXBackgroundFetch/EXBackgroundFetchTaskConsumer.m [hunk 5]
```diff
@@ -48,4 +54,14 @@ - (NSUInteger)normalizeTaskResult:(id)result
   }
 }
 
+- (void)updateMinimumInterval
+{
+  NSNumber *interval = _task.options[@"minimumInterval"];
+  NSTimeInterval timeInterval = interval != nil ? [interval doubleValue] : UIApplicationBackgroundFetchIntervalMinimum;
```
Comment (sjchmiela) [inline packages/expo-background-fetch/ios/EXBackgroundFetch/EXBackgroundFetchTaskConsumer.m]:
```suggestion
  NSTimeInterval timeInterval = [interval doubleValue] ?: UIApplicationBackgroundFetchIntervalMinimum;
```

**Review by sjchmiela — APPROVED** [review]
LGTM, just two minor comments

File: packages/expo-background-fetch/ios/EXBackgroundFetch/EXBackgroundFetchTaskConsumer.m (hunk 4 above)
Comment (tsapeta) [reply packages/expo-background-fetch/ios/EXBackgroundFetch/EXBackgroundFetchTaskConsumer.m]: No, we can access to them by `_task.options` if needed

File: docs/versions/unversioned/sdk/background-fetch.md [hunk 6]
```diff
@@ -17,13 +16,18 @@ Returns a promise resolving to one of these values:
...
+-   **taskName (_string_)** -- Name of the task to register. The task needs to be defined first - see [TaskManager.defineTask](../task-manager#taskmanagerdefinetasktaskname-task) for more details.
+-   **options (_object_)** -- An object of options:
+    -   **minimumInterval (_number_)** -- Inexact interval in seconds between subsequent repeats of the background fetch alarm. The final interval may differ from the specified one to minimize wakeups and battery usage.
+    On Android it defaults to **15 minutes**. On iOS it calls [BackgroundFetch.setMinimumIntervalAsync](#backgroundfetchsetminimumintervalasyncminimuminterval) behind the scenes and the default value is the smallest fetch interval supported by the system (**10-15 minutes**).
+    -   **stopOnTerminate (_boolean_)** -- Whether to stop receiving background fetch events after user terminates the app. Defaults to `true`. (**Android only**)
+    -   **startOnBoot (_boolean_)** -- Whether to restart background fetch events when the device has finished booting. Defaults to `false`. (**Android only**)
```
Comment (BrodaNoel) [inline docs/versions/unversioned/sdk/background-fetch.md]: Is there a way to add a bit more information about the `stopOnTerminate`?
What does exactly means "after user terminates the app"?

Comment (brentvatne) [reply docs/versions/unversioned/sdk/background-fetch.md]: background app: switch to another
terminate app: close it

**Pull request merged**